engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift

```diff
@@ -0,0 +1,116 @@
+/// They are ordered by increasing severity.
+@objc(FlutterLogLevel) public enum LogLevel: Int {
+  /// Informational messages that are helpful for tracing application flow.
+  case info = 0
+
+  /// Messages that highlight significant progress or state changes in the application.
+  case important
+
+  /// Messages indicating a potential issue or an unexpected situation that isn't critical.
+  case warning
+
+  /// Messages indicating a runtime error from which the application can potentially recover.
+  case error
+
+  /// Messages indicating a critical condition. Causes the application to immediately terminate.
+  case fatal
+}
+
+/// A singleton logger for outputting runtime messages.
+///
+/// This logger allows for messages to be logged at different severity levels. Its output can be
+/// filtered by setting the `logLevel` property to the minimum log level to be logged.
+///
+/// **Usage:**
+/// ```swift
+/// Logger.logInfo("Application has started.")
+/// Logger.instance.logLevel = .warning // Only show warnings and above
+/// Logger.logError("Failed to load asset: \(assetKey)")
+/// ```
+@objc(FlutterLogger) public final class Logger: NSObject {
+  /// The shared singleton instance of the logger.
+  @objc public static let instance = Logger()
+
+  /// Logs a message at `LogLevel.info`.
+  @objc public static func logInfo(_ message: String) {
+    instance.log(level: .info, message)
+  }
+
+  /// Logs a message at `LogLevel.important`.
+  @objc public static func logImportant(_ message: String) {
+    instance.log(level: .important, message)
+  }
+
+  /// Logs a message at `LogLevel.warning`.
+  @objc public static func logWarning(_ message: String) {
+    instance.log(level: .warning, message)
+  }
+
+  /// Logs a message at `LogLevel.error`.
+  @objc public static func logError(_ message: String) {
+    instance.log(level: .error, message)
+  }
+
+  /// Logs a message at `LogLevel.fatal` and immediately terminates the application.
+  @objc public static func logFatal(_ message: String) {
+    instance.log(level: .fatal, message)
+    abort()
+  }
+
+  /// Logs a message unconditionally.
+  @objc public static func logDirect(_ message: String) {
+    instance.outputWriter.writeLine(message)
+  }
+
+  /// Sets the minimum log level.
+  @objc public static func setLogLevel(_ level: LogLevel) {
+    instance.logLevel = level
+  }
+
+  var logLevel = LogLevel.info
+  let outputWriter: OutputWriter
+
+  private override init() {
+#if os(iOS)
+    outputWriter = SyslogOutputWriter()
+#elseif os(macOS)
+    outputWriter = StdoutOutputWriter()
+#endif
+  }
+
+  private func log(level: LogLevel, _ message: String) {
+    if level.rawValue >= logLevel.rawValue {
+      outputWriter.writeLine(message)
+    }
+  }
+}
+
+protocol OutputWriter {
+  func writeLine(_ message: String)
+}
+
+final class SyslogOutputWriter: OutputWriter {
+  func writeLine(_ message: String) {
+    // TODO(cbracken): replace this with os_log-based approach.
```

Comment:
I don't know how feasible this is, but I would love to replace this with proper string formatting, since static strings aren't redacted in logs, but dynamic ones are:
https://developer.apple.com/documentation/os/generating-log-messages-from-your-code?#Redact-Sensitive-User-Data-from-a-Log-Message

Reply:
Agreed -- this one tries to purposely keep things super stupid and port line for line what we do today, but once we have this all centralised here, then we absolutely should look at trying for os_log one more time and getting all the nice stuff like that!

I originally tried to export a varargs version of this but `@objc` bridging doesn't support it, sadly. We could make a Swift-only API alongside these though so at least Swift code would have it!